PowerShellTools/LanguageService/SmartIndent.cs

```diff
@@ -11,172 +11,170 @@ namespace PowerShellTools.LanguageService
-		textBuffer.Delete(new Span(line.Start, precedingWhiteSpaces));
-	    }
-
-	    return indentation;
-	}
-
-	private static bool FindGroupEndWithoutNonWhiteSpacesPreceding(ITextSnapshotLine line, char groupStartChar, out SnapshotPoint groupEnd)
-	{
-	    string lineText = line.GetText();
-	    int lineNumber = line.LineNumber;
-	    char groupEndChar = Utilities.GetPairedBrace(groupStartChar);
-	    groupEnd = new SnapshotPoint();
-
-	    //walk the entire line 
-	    for (int offset = 0; offset < line.Length; offset++)
-	    {
-		char currentChar = lineText[offset];
-		if (currentChar == groupEndChar)
-		{
-		    groupEnd = new SnapshotPoint(line.Snapshot, line.Start + offset);
-		    return true;
-		}
-		if (!char.IsWhiteSpace(currentChar))
-		{
-		    return false;
-		}
-	    }
-	    return false;
-	}
+        private ITextView _textView;
+        private PowerShellLanguageInfo _info;
+        private static readonly ILog Log = LogManager.GetLogger(typeof(SmartIndent));
+
+        /// <summary>
+        /// Constructor.
+        /// </summary>
+        /// <param name="info">Powershell language service.</param>
+        /// <param name="textView">Current active TextView.</param>
+        public SmartIndent(PowerShellLanguageInfo info, ITextView textView)
+        {
+            _info = info;
+            _textView = textView;
+        }
+
+        /// <summary>
+        /// Implementation of the interface.
+        /// </summary>
+        /// <param name="line">The current line after Enter.</param>
+        /// <returns>Desired indentation size.</returns>
+        public int? GetDesiredIndentation(ITextSnapshotLine line)
+        {
+            // User GetIndentSize() instead of GetTabSize() due to the fact VS always uses Indent Size as a TAB size
+            int tabSize = _textView.Options.GetIndentSize();
+
+            switch (_info.LangPrefs.IndentMode)
+            {
+                case vsIndentStyle.vsIndentStyleNone:
+                    return null;
+
+                case vsIndentStyle.vsIndentStyleDefault:
+                    return GetDefaultIndentationImp(line, tabSize);
+
+                case vsIndentStyle.vsIndentStyleSmart:
+                    return GetSmartIndentationImp(line, tabSize);
+            }
+            return null;
+        }
+
+        public void Dispose() { }
+
+        /// <summary>
+        /// Implementation of default indentation.
+        /// </summary>
+        /// <param name="line">The current line after Enter.</param>
+        /// <param name="tabSize">The TAB size.</param>
+        /// <returns>Desired indentation size.</returns>
+        private int? GetDefaultIndentationImp(ITextSnapshotLine line, int tabSize)
+        {
+            int lineNumber = line.LineNumber;
+            if (lineNumber < 1) return 0;
+
+            string baselineText = null;
+            ITextSnapshotLine baseline = null;
+            IndentUtilities.SkipPrecedingBlankLines(line, out baselineText, out baseline);

```

Comment:
What if there are only blank lines preceding this line, would the baselinetext end up null? I don't see GetCurrentLineIndentation check null before access .Length, which make me guess it would be an NullRefEx thrown?